frontend/src/components/Tree.tsx

```diff
@@ -0,0 +1,291 @@
+import { Fragment, useMemo, useState } from "react";
+import { curveStepBefore, hierarchy, line, type HierarchyNode } from "d3";
+import { map, mapValues, max, min, orderBy, sum, uniqueId } from "lodash";
+import Chart from "@/components/Chart";
+import Legend from "@/components/Legend";
+import Tooltip from "@/components/Tooltip";
+import { useColorMap } from "@/util/color";
+import type { Filename } from "@/util/download";
+import { useTheme, useTruncateWidth } from "@/util/hooks";
+import { round } from "@/util/math";
+import classes from "./Tree.module.css";
+
+/** grid spacing */
+const size = 30;
+/** circle size */
+const nodeSize = 6;
+/** line thickness */
+const lineWidth = 1;
+/** width of legend and labels */
+const sideWidth = 150;
+
+type Props = {
+  /** title text */
+  title?: string;
+  /** download filename */
+  filename?: Filename;
+  /** chart data */
+  data: Item[];
+};
+
+export type Item = {
+  /** human-readable label */
+  label?: string;
+  /** arbitrary type/category */
+  type?: string;
+  /** distance from parent */
+  dist?: number;
+  /** children items */
+  children?: Item[];
+};
+
+/** link line generator */
+const link = line().curve(curveStepBefore);
+
+/** tree/hierarchy plot */
+const Tree = ({ title, filename = [], data }: Props) => {
```

Comment:
This is not a new file, it was added in #54, so much of this diff will go away once the tree PR is merged and I rebase this PR.